File: Task2/ex2Python/morse.py
# Таблица кодов Морзе
MORSE_CODE = {
    'a': ".-", 'b': "-...", 'c': "-.-.", 'd': "-..", 'e': ".", 'f': "..-.",
    'g': "--.", 'h': "....", 'i': "..", 'j': ".---", 'k': "-.-", 'l': ".-..",
    'm': "--", 'n': "-.", 'o': "---", 'p': ".--.", 'q': "--.-", 'r': ".-.",
    's': "...", 't': "-", 'u': "..-", 'v': "...-", 'w': ".--", 'x': "-..-",
    'y': "-.--", 'z': "--.."
}

def word_to_morse(word):
    """Преобразует слово в код Морзе с проверкой символов"""
    morse = []
    for char in word:
        if char not in MORSE_CODE:
            raise ValueError(f"Недопустимый символ: {char}")
        morse.append(MORSE_CODE[char])
    return ''.join(morse)

def count_unique_morse(words):
    """Подсчитывает уникальные коды Морзе с учетом перестановок букв"""
    unique_codes = set()
    
    for word in words:
        # Проверка длины слова
        if len(word) > 12:
            print(f"Ошибка: слово '{word}' слишком длинное (максимум 12 символов)")
            return None
        
        try:
            morse = word_to_morse(word)
            # Сортируем символы для учета перестановок
            sorted_morse = ''.join(sorted(morse))
            unique_codes.add(sorted_morse)
        except ValueError as e:
            print(f"Ошибка: {e}")
            return None
    
    return len(unique_codes)

def main():
    print("Введите слова через пробел:")
    try:
        input_words = input().strip().split()
        
        # Проверка количества слов
        if len(input_words) > 100:
            print("Ошибка: слишком много слов (максимум 100)")
            return
        
        result = count_unique_morse(input_words)
        if result is not None:
            print(f"Количество уникальных кодов Морзе: {result}")
    except Exception as e:
        print(f"Произошла ошибка: {e}")

if __name__ == "__main__":
    main()

File: Task2/ex2Ruby/morse.rb
# Таблица кодов Морзе
MORSE_CODE = {
  'a' => ".-", 'b' => "-...", 'c' => "-.-.", 'd' => "-..", 'e' => ".", 'f' => "..-.",
  'g' => "--.", 'h' => "....", 'i' => "..", 'j' => ".---", 'k' => "-.-", 'l' => ".-..",
  'm' => "--", 'n' => "-.", 'o' => "---", 'p' => ".--.", 'q' => "--.-", 'r' => ".-.",
  's' => "...", 't' => "-", 'u' => "..-", 'v' => "...-", 'w' => ".--", 'x' => "-..-",
  'y' => "-.--", 'z' => "--.."
}.freeze

# Преобразует слово в код Морзе
def word_to_morse(word)
  word.chars.map do |char|
    MORSE_CODE[char] || raise("Недопустимый символ: #{char}")
  end.join
end

# Подсчитывает уникальные коды Морзе с учетом перестановок букв
def count_unique_morse(words)
  unique_codes = Set.new
  
  words.each do |word|
    # Проверка длины слова
    if word.length > 12
      puts "Ошибка: слово '#{word}' слишком длинное (максимум 12 символов)"
      return nil
    end

    begin
      morse = word_to_morse(word)
      # Сортируем символы для учета перестановок
      sorted_morse = morse.chars.sort.join
      unique_codes.add(sorted_morse)
    rescue => e
      puts "Ошибка: #{e.message}"
      return nil
    end
  end
  
  unique_codes.size
end

# Основная функция
def main
  puts "Введите слова через пробел:"
  input = gets.chomp
  words = input.split
  
  # Проверка количества слов
  if words.size > 100
    puts "Ошибка: слишком много слов (максимум 100)"
    return
  end
  
  result = count_unique_morse(words)
  puts "Количество уникальных кодов Морзе: #{result}" if result
end

# Запуск программы
main if __FILE__ == $0